OGM-732 Queries with entities having InheritanceType.SINGLE_TABLE

DavideD opened the pull request:
https://hibernate.atlassian.net/browse/OGM-732

This patch make it possible to run queries on entity hierarchies when mapped with InheritanceType.SINGLE_CLASS. This wasn't working before for Neo4j and MongoDB.

I haven't implemented the TABLE_PER_CLASS approach, so , for now, it throws an exception

See also https://github.com/hibernate/hibernate-ogm/pull/803 for branch 5.0

gsmet commented on neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/query/parsing/CypherQueryParsingTest.java:
```diff
@@ -0,0 +1,273 @@
+import org.junit.Test;
+
+/**
+ * Integration test for {@link org.hibernate.ogm.datastore.mongodb.query.parsing.impl.MongoDBQueryResolverDelegate} and
```
It's not related to MongoDB!

gsmet commented on neo4j/src/test/java/org/hibernate/ogm/datastore/neo4j/test/query/parsing/CypherQueryParsingTest.java:
```diff
@@ -0,0 +1,273 @@
+ * Integration test for {@link org.hibernate.ogm.datastore.mongodb.query.parsing.impl.MongoDBQueryResolverDelegate} and
+ * {@link org.hibernate.ogm.datastore.mongodb.query.parsing.impl.MongoDBQueryRendererDelegate}.
+ *
+ * @author Gunnar Morling
```
s/Gunnar Morling/Davide D'Alto/?
(looks like you rewrote everything)

DavideD replied:
I should stop using  copy and paste :)

gsmet commented on core/src/main/java/org/hibernate/ogm/util/impl/Log.java:
```diff
@@ -301,4 +302,6 @@
 	@Message(id = 88, value = "Configuration is referring to deprecated datastore provider name '%1$s'. Please use the new form '%2$s' instead.")
 	void usingDeprecatedDatastoreProviderName(String deprecatedName, String newName);
 
+	@Message(id = 89, value = "%1$s does not support queries on polymorfic entities using TABLE_PER_CLASS inheritance strategy. You should try using SINGLE_TABLE instead. Entities: %2$s")
```
s/polymorfic/polymorphic/

DavideD commented on core/src/main/java/org/hibernate/ogm/persister/impl/SingleTableOgmEntityPersister.java:
```diff
@@ -37,6 +38,10 @@ public SingleTableOgmEntityPersister(
 		if ( persistentClass.isPolymorphic() ) {
 			Value discrimValue = persistentClass.getDiscriminator();
+			if ( discrimValue == null ) {
+				throw new MappingException( "discriminator mapping required for single table polymorphic persistence" );
```
I copied this from Hibernate ORM, not sure how much we want to they similar to what they do

DavideD commented on mongodb/src/test/java/org/hibernate/ogm/datastore/mongodb/test/inheritance/SingleTableInheritanceTest.java:
```diff
@@ -0,0 +1,130 @@
+		try {
+			em.getTransaction().begin();
+			@SuppressWarnings("unchecked")
+			List<TextNode> resultList = em.createNativeQuery( "db.Node.find({'DTYPE': 'TextNode'})", TextNode.class ).getResultList();
```
Well, the expected return type is a parameter of the createNativeQuery method, so yes.

I'm not sure what happens if one doesn't specified the returned type. I think it returns an array of objects

gsmet replied:
Yes, my remark was more to outline that when using the DTYPE query directly, we don't return the possible subclasses. I think we can't do better but I wanted to be sure I understood the limitations right.

DavideD commented:
I push-forced all the changes and removed the confusing part in OgmLoader.
It wasn't needed and the main goal was to align the behaviour to ORM. I will keep it for another pull request.

I think now, the patch is much simpler and we can actually apply it to the 5.0 branch.
It seems odd that people cannot execute queries for entities using SINGLE_TABLE inheritance strategy.

I think we have implemented all the features needed for a final release. My plan now is to release a Beta2, align class names and improve the documentation if needed and then make another release before a CR1 and a Final.

Let's see how it goes, we received quite some feedback recently and I hope the trend will continue.

gsmet commented:
Nice work! Merged!

I'm wondering if we should add some documentation on what's supported or not depending on the backend. I remember you updating the doc on inheritance recently. Maybe a table with what's supported by each backend would be in order? WDYT?

DavideD commented:
Yes, I need to update the documentation on several front. I was waiting for this PR to be integrated.
I will update the doc working on https://hibernate.atlassian.net/browse/OGM-1216

gsmet commented:
OK cool, glad we're on the same page! Closing then.
